skimage/filters/thresholding.py

```diff
@@ -970,6 +970,106 @@ def _mean_std(image, w):
+    The algorithm uses integral images to speedup computation.
+
+    Parameters
+    ----------
+    image : ndarray
+        Input image.
+    w : int, or iterable of int
+        Window size specified as a single odd integer (3, 5, 7, …),
+        or an iterable of length ``image.ndim`` containing only odd
+        integers (e.g. ``(1, 5, 5)``).
+
+    Returns
+    -------
+    m : ndarray of float, same shape as ``image``
+        Local mean of the image.
+    """
+
+   if not isinstance(w, Iterable):
+       w = (w,) * image.ndim
+   _validate_window_size(w)
+
+   pad_width = tuple((k // 2 + 1, k // 2) for k in w)
+   padded = np.pad(image.astype('float'), pad_width,
+                    mode='reflect')
+   #padded_sq = padded * padded
+
+   integral = integral_image(padded)
+   #integral_sq = integral_image(padded_sq)
+
+   kern = np.zeros(tuple(k + 1 for k in w))
+   for indices in itertools.product(*([[0, -1]] * image.ndim)):
+       kern[indices] = (-1) ** (image.ndim % 2 != np.sum(indices) % 2)
+
+   total_window_size = np.prod(w)
+   sum_full = ndi.correlate(integral, kern, mode='constant')
+   m = crop(sum_full, pad_width) / total_window_size
+   #sum_sq_full = ndi.correlate(integral_sq, kern, mode='constant')
+   #g2 = crop(sum_sq_full, pad_width) / total_window_size
+   # Note: we use np.clip because g2 is not guaranteed to be greater than
+   # m*m when floating point error is considered
+   #s = np.sqrt(np.clip(g2 - m * m, 0, None))
+   return m#, s
+  
+def threshold_singh(image, window_size=15, k=0.2, r=None):
+    """Applies Singh local threshold to an array.
+
+    The threshold T is calculated for every pixel in the image using
+    the following formula:
+
+        T = m(x,y) * (1 + k * ((d(x,y) / 1 - d(x,y)) - 1))
+
+    where m(x,y) and d(x,y) are the local mean and the local deviation of
+    pixel (x,y) neighborhood defined by a rectangular window with size wxw
+    centered around the pixel. k is a configurable parameter
+    that weights the effect of standard deviation.
+   
+    Parameters
+    ----------
+    image : ndarray
+        Input image.
+    window_size : int, or iterable of int, optional
+        Window size specified as a single odd integer (3, 5, 7, …),
+        or an iterable of length ``image.ndim`` containing only odd
+        integers (e.g. ``(1, 5, 5)``).
+    k : float, optional
+        Value of the positive parameter k.
+
+    Returns
+    -------
+    threshold : (N, M) ndarray
+        Threshold mask. All pixels with an intensity higher than
+        this value are assumed to be foreground.
+
+    Notes
+    -----
+    This algorithm is originally designed for text recognition.
+
+    References
+    ----------
+    [1] T Romen Singh, Sudipta Roy, O Imocha Singh, Tejmani Sinam, Kh Manglem Singh.
+        "A New Local Adaptive Thresholding Technique in Binarization." IJCSI
+        International Journal of Computer Science Issues. 2011; 8(6-2): 271-276.
+
```

Comment:
```suggestion
        International Journal of Computer Science Issues. 2011; 8(6-2): 271-276.
        http://ijcsi.org/papers/IJCSI-8-6-3-275-280.pdf
```

We may as well also provide the URL to the (freely available) publication. As far as I could tell, there does not appear to be an associated DOI.